driver/bpf/fillers.h

```diff
@@ -571,236 +571,158 @@ FILLER(sys_poll_x, true)
 	return bpf_poll_parse_fds(data, false);
 }
 
-#define MAX_IOVCNT 32
-#define MAX_IOVCNT_COMPAT 8
+#ifdef CONFIG_COMPAT
+	#define MAX_IOVCNT 8
+#else
+	#define MAX_IOVCNT 32
+#endif
 
-static __always_inline int bpf_parse_readv_writev_bufs_64(struct filler_data *data,
+static __always_inline int bpf_parse_readv_writev_bufs(struct filler_data *data,
 						       const void __user *iovsrc,
 						       unsigned long iovcnt,
 						       long retval,
-						       int flags,
-						       unsigned long *size)
+						       int flags)
 {
-	const struct iovec *iov;
 	int res = PPM_SUCCESS;
 	unsigned long copylen;
+	long size = 0;
 	int j;
+	unsigned long iov_size = sizeof(struct iovec);
+	unsigned long len_off = offsetof(struct iovec, iov_len);
+	unsigned long base_off = offsetof(struct iovec, iov_base);
+	unsigned long ptr_size = sizeof(void *);
 
-	copylen = iovcnt * sizeof(struct iovec);
-	iov = (const struct iovec *)data->tmp_scratch;
+#ifdef CONFIG_COMPAT
+	if (bpf_in_ia32_syscall())
+	{
+		iov_size = sizeof(struct compat_iovec);
+		len_off = offsetof(struct compat_iovec, iov_len);
+		base_off = offsetof(struct compat_iovec, iov_base);
+		ptr_size = 4;
+	}
+#endif
 
+	copylen = iovcnt * iov_size;
 	if (copylen > SCRATCH_SIZE_MAX)
 	{
 		return PPM_FAILURE_FRAME_SCRATCH_MAP_FULL;
 	}
 
 #ifdef BPF_FORBIDS_ZERO_ACCESS
-		if (copylen)
-			if (bpf_probe_read_user((void *)iov,
-						((copylen - 1) & SCRATCH_SIZE_MAX) + 1,
-						(void *)iovsrc))
+	if (copylen)
+		if (bpf_probe_read_user(data->tmp_scratch,
+				       ((copylen - 1) & SCRATCH_SIZE_MAX) + 1,
+				       (void *)iovsrc))
 #else
-		if (bpf_probe_read_user((void *)iov,
-					copylen & SCRATCH_SIZE_MAX,
-					(void *)iovsrc))
+	if (bpf_probe_read_user(data->tmp_scratch,
+			       copylen & SCRATCH_SIZE_MAX,
+			       (void *)iovsrc))
 #endif
 			return PPM_FAILURE_INVALID_USER_MEMORY;
 
-	#pragma unroll
+
+#pragma unroll
 	for (j = 0; j < MAX_IOVCNT; ++j) {
 		if (j == iovcnt)
 			break;
 		// BPF seems to require a hard limit to avoid overflows
-		if (*size == LONG_MAX)
+		if (size == LONG_MAX)
 			break;
 
-		*size += iov[j].iov_len;
+		volatile unsigned curr_shift = j * iov_size + len_off;
+		unsigned long shift_bounded = curr_shift & SCRATCH_SIZE_HALF;
+		if (curr_shift > SCRATCH_SIZE_HALF)
+			break;
+
+		long curr_len;
```

Comment:
Don't know why, it won't make me use `memcpy` (that is an alias for `__builtin_memcpy` from clang) here.

Reply:
If i could use `memcpy` the patch would look much tidier, but that's not the case unfortunately.

Reply:
Oh we can't because for `__builtin_memcpy` size must be a compile time constant.